chained/water_impl.go

```diff
@@ -115,11 +129,31 @@ func createDialer(ctx context.Context, wasm []byte, transport string) (water.Dia
 	return dialer, nil
 }
 
+func (d *waterImpl) isReady() bool {
+	d.readyMutex.Lock()
+	defer d.readyMutex.Unlock()
+	return d.ready
+}
+
+func (d *waterImpl) close() {
+	close(d.dialerAvailable)
+}
+
 func (d *waterImpl) dialServer(op *ops.Op, ctx context.Context) (net.Conn, error) {
 	return d.reportDialCore(op, func() (net.Conn, error) {
-		d.wgDownload.Wait()
-		if d.dialer == nil {
-			return nil, log.Errorf("dialer not available: %w", d.downloadErr)
+		// if dialer is not ready, wait until WASM is downloaded or context timeout
+		if !d.isReady() {
+			select {
+			case <-d.dialerAvailable:
+				log.Debug("download finished!")
+				// carry on with dial
```

Comment:
You missed the `ok` check here so this doesn't continue to dial if `close` is called while it's waiting.